src/pip/_internal/operations/prepare.py

```diff
@@ -536,9 +552,38 @@ def _prepare_linked_requirement(
         assert req.link
         link = req.link
 
-        self._ensure_link_req_src_dir(req, parallel_builds)
         hashes = self._get_linked_req_hashes(req)
 
+        if hashes and req.original_link_is_in_wheel_cache:
+            assert req.download_info is not None
+            assert link.is_wheel
+            assert link.is_file
+            # We need to verify hashes, and we have found the requirement in the cache
+            # of locally built wheels.
+            if (
+                isinstance(req.download_info.info, ArchiveInfo)
+                and req.download_info.info.hashes
+                and hashes.has_one_of(req.download_info.info.hashes)
+            ):
+                # At this point we know the requirement was built from a hashable source
+                # artifact, and we verified that the cache entry's hash of the original
+                # artifact matches one of the hashes we expect. We don't verify hashes
+                # against the cached wheel, because the wheel is not the original.
+                hashes = None
+            else:
+                # Re-download the original link found in download_info.
+                # We re-download instead of failing because the cache entry may have
+                # hashes with different algorithms than those expected by the user.
+                logger.warning(
+                    "The hashes of the source archive found in cache entry "
+                    "don't match, ignoring cached built wheel "
+                    "and re-downloading source."
+                )
+                req.link = direct_url_as_link(req.download_info)
+                link = req.link
```

Comment:
Hmm, I'm afraid this part still needs some work. If req.download_info contains an invalid hash, it re-downloads an URL with an invalid hash fragment, which means we never get a good cache entry.

Reply:
Ok, I resolved this by adding a new `cached_wheel_source_link`  field to `InstallRequirement`. I also renamed `original_link_is_in_wheel_cache` to `is_wheel_from_cache` to better reflect its meaning.